pydantic/_internal/_model_construction.py

```diff
@@ -166,9 +166,7 @@ def hash_func(self: Any) -> int:
             # preserve `__set_name__` protocol defined in https://peps.python.org/pep-0487
             # for attributes not in `new_namespace` (e.g. private attributes)
             for name, obj in private_attributes.items():
-                set_name = getattr(obj, '__set_name__', None)
-                if callable(set_name):
-                    set_name(cls, name)
+                obj.__set_name__(cls, name)
```

Comment:
We deliberately construct `private_attributes` as `dict[str, PrivateAttribute]` and `PrivateAttribute` has this method with the same hack already.